src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,290 @@
+package duke;
+
+import duke.exception.ParseException;
+import duke.exception.InvalidCommandException;
+import duke.exception.TaskIndexNotSpecifiedException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDo;
+
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.util.ArrayList;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+import static duke.Storage.FILE_PATH;
+import static duke.Ui.BREAK;
+import static duke.Ui.COMMAND_TODO;
+import static duke.Ui.COMMAND_EVENT;
+import static duke.Ui.COMMAND_DEADLINE;
+import static duke.Ui.COMMAND_DONE;
+import static duke.Ui.COMMAND_DELETE;
+
+/**
+ * TaskList stores Tasks in an ArrayList of Task objects, and handles any and all bookkeeping of Tasks.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
+    private Storage storage;
+    private Parser parser;
+    private final static int STARTING_TASK_NUMBER = 1;
+
+    /**
+     * Constructor. Initializes a new ArrayList and new Storage. Loads the TaskList from a saved file, if any.
+     *
+     * @throws FileNotFoundException
+     */
+    public TaskList() throws FileNotFoundException {
+        tasks = new ArrayList<>();
+        storage = new Storage();
+        parser = new Parser();
+        loadTaskList();
+    }
+
+    /**
+     * Constructor. Creates a TaskList from an existing List of Tasks.
+     *
+     * @param tasks
+     */
+    public TaskList(List<Task> tasks) {
+        this.tasks = (ArrayList<Task>) tasks;
+    }
+
+    /**
+     * Reads saved file data into TaskList. Catches a ParseException for errors encountered while reading data.
+     *
+     * @throws FileNotFoundException
+     */
+    public void loadTaskList() throws FileNotFoundException {
+        File saveFile = storage.load();
+        if (saveFile.exists()) {
```

Comment:
Potential arrow head code, perhaps can restructure this part to make the happy path unindented as much as possible. It is the ‘unusual’ cases that should be indented. One technique that could help in this regard is the use of guard clauses.

Reply:
Implemented (with guard clause)! Thanks for suggestion!